src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php

```diff
@@ -182,6 +167,130 @@ private function getEventFromTypeDeclaration(ContainerBuilder $container, string
+            foreach ($listenersForEvent as $listenerName => $listenerDefinition) {
+                if (isset($listenerDefinition['before']) || isset($listenerDefinition['after'])) {
+                    if ($listenerDefinition['before'] && !isset($listenersForEvent[$listenerDefinition['before']])) {
+                        throw new InvalidArgumentException(sprintf('Given "before" "%s" for listener "%s" does not listen the same event.', $listenerDefinition['before'], $listenerName));
+                    }
+
+                    if ($listenerDefinition['after'] && !isset($listenersForEvent[$listenerDefinition['after']])) {
+                        throw new InvalidArgumentException(sprintf('Given "after" "%s" for listener "%s" does not listen the same event.', $listenerDefinition['after'], $listenerName));
+                    }
+
+                    $this->updateListenerPriority(
+                        $container,
+                        $eventName,
+                        $listenerName,
+                        $this->computePriority($listenersForEvent, $listenerName)
+                    );
+                }
+            }
+        }
+    }
+
+    private function guessEventAndMethod(array &$eventDefinition, ContainerBuilder $container, string $serviceId, array $aliases): void
+    {
+        if (!isset($eventDefinition['event'])) {
+            $eventDefinition['method'] ??= '__invoke';
+            $eventDefinition['event'] = $this->getEventFromTypeDeclaration($container, $serviceId, $eventDefinition['method']);
+        }
+
+        $eventDefinition['event'] = $aliases[$eventDefinition['event']] ?? $eventDefinition['event'];
+
+        if (!isset($eventDefinition['method'])) {
+            $eventDefinition['method'] = 'on'.preg_replace_callback([
+                    '/(?<=\b|_)[a-z]/i',
+                    '/[^a-z0-9]/i',
+                ], fn ($matches) => strtoupper($matches[0]), $eventDefinition['event']);
+            $eventDefinition['method'] = preg_replace('/[^a-z0-9]/i', '', $eventDefinition['method']);
+
+            if (null !== ($class = $container->getDefinition($serviceId)->getClass()) && ($r = $container->getReflectionClass($class, false)) && !$r->hasMethod($eventDefinition['method']) && $r->hasMethod('__invoke')) {
+                $eventDefinition['method'] = '__invoke';
+            }
+        }
+    }
+
+    private function normalizeBeforeAfter(string|null $beforeAfter, array $aliases): string|null
+    {
+        if (null === $beforeAfter) {
+            return null;
+        }
+
+        $beforeAfter = explode('::', $beforeAfter);
+
+        return sprintf(
+            '%s::%s',
+            $aliases[$beforeAfter[0]] ?? $beforeAfter[0],
+            trim($beforeAfter[1] ?? '__invoke', '()')
+        );
+    }
+
+    private function computePriority(array $listeners, string $listenerName, array $alreadyVisited = []): int
+    {
+        if ($alreadyVisited[$listenerName] ?? false) {
```

Comment:
```suggestion
        if (isset($alreadyVisited[$listenerName]) {
```
Might be faster

Reply:
ok but it is such a micro optim! :sweat_smile: 
https://3v4l.org/sY5ZR#v8.2.7

Reply:
Yet more explicit IMHO